frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts

```diff
@@ -0,0 +1,681 @@
+    },
+    reducers: {
+        isPaletteShown: [
+            false,
+            {
+                hidePalette: () => false,
+                showPalette: () => true,
+                togglePalette: (previousState) => !previousState,
+            },
+        ],
+        keyboardResultIndex: [
+            0,
+            {
+                setInput: () => 0,
+                executeResult: () => 0,
+                activateFlow: () => 0,
+                onArrowUp: (previousIndex) => (previousIndex > 0 ? previousIndex - 1 : 0),
+                onArrowDown: (previousIndex, { maxIndex }) => (previousIndex < maxIndex ? previousIndex + 1 : maxIndex),
+            },
+        ],
+        hoverResultIndex: [
+            null as number | null,
+            {
+                onMouseEnterResult: (_, { index }) => index,
+                onMouseLeaveResult: () => null,
+                onArrowUp: () => null,
+                onArrowDown: () => null,
+                activateFlow: () => null,
+            },
+        ],
+        input: [
+            '',
+            {
+                setInput: (_, { input }) => input,
+                activateFlow: () => '',
+                executeResult: () => '',
+            },
+        ],
+        activeFlow: [
+            null as CommandFlow | null,
+            {
+                activateFlow: (_, { flow }) => flow,
+            },
+        ],
+        rawCommandRegistrations: [
+            {} as CommandRegistrations,
+            {
+                registerCommand: (commands, { command }) => {
+                    return { ...commands, [command.key]: command }
+                },
+                deregisterCommand: (commands, { commandKey }) => {
+                    const cleanedCommands = { ...commands }
+                    delete cleanedCommands[commandKey]
+                    return cleanedCommands
+                },
+            },
+        ],
+    },
+
+    listeners: ({ actions, values }) => ({
+        showPalette: () => {
+            window.posthog?.capture('palette shown')
+        },
+        togglePalette: () => {
+            if (values.isPaletteShown) window.posthog?.capture('palette shown')
+        },
+        executeResult: ({ result }: { result: CommandResult }) => {
+            if (result.executor === true) {
+                actions.activateFlow(null)
+            } else {
+                const possibleFlow = result.executor?.() ?? null
+                actions.activateFlow(possibleFlow)
+                if (!possibleFlow) actions.hidePalette()
+            }
+            // Capture command execution, without useless data
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            const { icon, index, ...cleanedResult } = result
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            const { resolver, ...cleanedCommand } = cleanedResult.source
+            cleanedResult.source = cleanedCommand
```

Comment:
this gives a TS error

Reply:
It does, but we're not strict about TS at the moment and I don't know what to do about this error (without adding a lot of boilerplate).

Reply:
Doesn't seem too complex?

![image](https://user-images.githubusercontent.com/53387/95864500-c0a79c00-0d65-11eb-8cc4-1dc3eb1eb5a0.png)

If you're not doing anything with this other than sending to posthog.capture, just cast it as a `Record<string, any>` and I think we're good to go.

We're not too strict about TS because of all the broken old/legacy code that doesn't work well with it... plus libraries that are a bit behind the times. I would love to enable stricter TS checks and definitely don't want to come back and revisit this issue then. I believe all new code should be good TS to the best of our abilities.